commit d0475f3e1834d85357772088c6d39f0097ec9b20
Author: Chris Friedt <cfriedt@tenstorrent.com>
Date:   Thu Nov 20 13:13:50 2025 -0500

    lib: minimal + picolibc: required changes to support strfmon()
    
    Required changes to support strfmon()
    
    Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

diff --git a/lib/libc/minimal/CMakeLists.txt b/lib/libc/minimal/CMakeLists.txt
index ce5438007c9..efca233cb4b 100644
--- a/lib/libc/minimal/CMakeLists.txt
+++ b/lib/libc/minimal/CMakeLists.txt
@@ -15,6 +15,7 @@ zephyr_library_compile_options($<TARGET_PROPERTY:compiler,no_builtin>)
 zephyr_library_compile_options($<TARGET_PROPERTY:compiler,prohibit_lto>)
 
 zephyr_library_sources(
+  source/locale/locale.c
   source/stdlib/atoi.c
   source/stdlib/strtol.c
   source/stdlib/strtoul.c
diff --git a/lib/libc/minimal/include/locale.h b/lib/libc/minimal/include/locale.h
new file mode 100644
index 00000000000..9a16a00bb01
--- /dev/null
+++ b/lib/libc/minimal/include/locale.h
@@ -0,0 +1,57 @@
+/*
+ * SPDX-FileCopyrightText: Copyright The Zephyr Project Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_LIB_LIBC_MINIMAL_INCLUDE_LOCALE_H_
+#define ZEPHYR_LIB_LIBC_MINIMAL_INCLUDE_LOCALE_H_
+
+#include <stddef.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct lconv
+{
+    char* decimal_point; /* "." */
+    char* thousands_sep; /* "" */
+    char* grouping;             /* "" */
+    char* mon_decimal_point;    /* "" */
+    char* mon_thousands_sep;    /* "" */
+    char* mon_grouping;         /* "" */
+    char* positive_sign;        /* "" */
+    char* negative_sign;        /* "" */
+    char* currency_symbol;      /* "" */
+    char  frac_digits;          /* CHAR_MAX */
+    char  p_cs_precedes;        /* CHAR_MAX */
+    char  n_cs_precedes;        /* CHAR_MAX */
+    char  p_sep_by_space;       /* CHAR_MAX */
+    char  n_sep_by_space;       /* CHAR_MAX */
+    char  p_sign_posn;          /* CHAR_MAX */
+    char  n_sign_posn;          /* CHAR_MAX */
+    char* int_curr_symbol;      /* "" */
+    char  int_frac_digits;      /* CHAR_MAX */
+    char  int_p_cs_precedes;    /* CHAR_MAX */
+    char  int_n_cs_precedes;    /* CHAR_MAX */
+    char  int_p_sep_by_space;   /* CHAR_MAX */
+    char  int_n_sep_by_space;   /* CHAR_MAX */
+    char  int_p_sign_posn;      /* CHAR_MAX */
+    char  int_n_sign_posn;      /* CHAR_MAX */
+};
+
+char* setlocale(int category, const char* locale);
+struct lconv* localeconv(void);
+
+#define LC_ALL	    0
+#define LC_COLLATE  1
+#define LC_CTYPE    2
+#define LC_MONETARY 3
+#define LC_NUMERIC  4
+#define LC_TIME     5
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif /* ZEPHYR_LIB_LIBC_MINIMAL_INCLUDE_LOCALE_H_ */
diff --git a/lib/libc/minimal/source/locale/locale.c b/lib/libc/minimal/source/locale/locale.c
new file mode 100644
index 00000000000..b4c5dc0c5dd
--- /dev/null
+++ b/lib/libc/minimal/source/locale/locale.c
@@ -0,0 +1,58 @@
+/*
+ * SPDX-FileCopyrightText: Copyright The Zephyr Project Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <limits.h>
+#include <locale.h>
+
+#include <zephyr/toolchain.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+static const struct lconv C_lconv = {
+    .decimal_point = ".",
+    .thousands_sep = "",
+    .grouping = "",
+    .mon_decimal_point = "",
+    .mon_thousands_sep = "",
+    .mon_grouping = "",
+    .positive_sign = "",
+    .negative_sign = "",
+    .currency_symbol = "",
+    .frac_digits = CHAR_MAX,
+    .p_cs_precedes = CHAR_MAX,
+    .n_cs_precedes = CHAR_MAX,
+    .p_sep_by_space = CHAR_MAX,
+    .n_sep_by_space = CHAR_MAX,
+    .p_sign_posn = CHAR_MAX,
+    .n_sign_posn = CHAR_MAX,
+    .int_curr_symbol = "",
+    .int_frac_digits = CHAR_MAX,
+    .int_p_cs_precedes = CHAR_MAX,
+    .int_n_cs_precedes = CHAR_MAX,
+    .int_p_sep_by_space = CHAR_MAX,
+    .int_n_sep_by_space = CHAR_MAX,
+    .int_p_sign_posn = CHAR_MAX,
+    .int_n_sign_posn = CHAR_MAX,
+};
+
+/* only "C" locale is supported */
+char* setlocale(int category, const char* locale)
+{
+	ARG_UNUSED(category);
+
+	if ((locale == NULL) || ((locale[0] == 'C') && (locale[1] == '\0'))) {
+		return "C";
+	}
+
+	return NULL;
+}
+
+struct lconv* localeconv(void)
+{
+	return (struct lconv*)&C_lconv;
+}
diff --git a/lib/libc/picolibc/include/monetary.h b/lib/libc/picolibc/include/monetary.h
new file mode 100644
index 00000000000..ba3e494a0ba
--- /dev/null
+++ b/lib/libc/picolibc/include/monetary.h
@@ -0,0 +1,12 @@
+/*
+ * Copyright The Zephyr Project Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_MONETARY_H_
+#define ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_MONETARY_H_
+
+#include <zephyr/posix/monetary.h>
+
+#endif /* ZEPHYR_LIB_LIBC_PICOLIBC_INCLUDE_MONETARY_H_ */
diff --git a/lib/libc/newlib/include/monetary.h b/lib/libc/newlib/include/monetary.h
new file mode 100644
index 00000000000..f3a2cf4bc74
--- /dev/null
+++ b/lib/libc/newlib/include/monetary.h
@@ -0,0 +1,12 @@
+/*
+ * Copyright The Zephyr Project Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_MONETARY_H_
+#define ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_MONETARY_H_
+
+#include <zephyr/posix/monetary.h>
+
+#endif /* ZEPHYR_LIB_LIBC_NEWLIB_INCLUDE_MONETARY_H_ */
